### PR TITLE
Add CAST(decimal as integral/floating point)

### PR DESCRIPTION
**Pull request by ebeworld**
Adding casting from decimal to float, bigint, int, smallint, and tinyint in addition to existing conversion to double. 
- Added corresponding tests
- Refactored a little bit.

### CODE REVIEW

**Comment by ebeworld** [comment]
@mbasmanova @majetideepak would you take a look into this one?

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:549]: std::function, but looks like this can be a regular function that returns the result, no?

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:563]: This can be removed as this place is unreachable.

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:133]: What if toType is some other kind of integer, SMALLINT or TINYINT?

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:145]: Is this needed?

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:142]: Move this outside of the loop.

**Review by mbasmanova — COMMENTED** [review]
@ebeworld Some comments. Would you benchmark this cast to check its performance?

File: velox/expression/CastExpr.cpp
Comment (majetideepak) [reply velox/expression/CastExpr.cpp:563]: We should be throwing an unsupported error here for the toType?

File: velox/expression/CastExpr.cpp
Comment (majetideepak) [reply velox/expression/CastExpr.cpp:549]: I agree with Masha. We could rename this function and return the result instead.

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:894]: Why leave it as a TODO?

File: velox/type/Conversions.h
Comment (majetideepak) [inline velox/type/Conversions.h:292]: Do we need the TRUNCATE part? We don't take this path in the current implementation.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:133]: to unblock internal tests only supports int32.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:563]: @mbasmanova That place is reachable for types such as float, short etc?

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:549]: I will change accordingly.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:563]: @majetideepak thought null access would do it, when it is called. but you are right, unsupported is much better exception.

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [reply velox/expression/CastExpr.cpp:133]: Would you add support for all integers for completeness?

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:145]: It will throw out of range exceptions. Following patterns in applyCastPrimitivesDispatch.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:142]: done.

File: velox/expression/tests/CastExprTest.cpp
Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp:894]: Removed.

File: velox/type/Conversions.h
Comment (ebeworld) [reply velox/type/Conversions.h:292]: Agree. I put them to make it uniform with existing codes above.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:133]: On it.

**Comment by ebeworld** [comment]
@mbasmanova Any benchmark in your mind?

**Comment by ebeworld** [comment]
> @ebeworld Some comments. Would you benchmark this cast to check its performance?

@mbasmanova Any benchmark in your mind?

File: velox/type/Conversions.h
Comment (mbasmanova) [reply velox/type/Conversions.h:292]: It might be better not to include code that's not used. Do we need this 'cast' method? Can we have the caller call `folly::to<T>(v)` directly?

Otherwise, let's add some tests for this method.

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:177]: Thanks.

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:206]: Should we also support REAL type?

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:156]: This needs to be a USER error.

Would you specify that value is out of bounds in the error message to help debugging?

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:159]: What does `util::Converter<ToKind, void, false>::cast` do here?

File: velox/expression/tests/CastExprTest.cpp
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:229]: It would be nice to include the 'to' type in the error message and the reason for failure.

**Review by mbasmanova — COMMENTED** [review]
@ebeworld Looks great % a few comments.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:159]: Followed the same pattern in the code. Removed.

File: velox/type/Conversions.h
Comment (ebeworld) [reply velox/type/Conversions.h:292]: done.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:156]: Done.

File: velox/expression/tests/CastExprTest.cpp
Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp:229]: Done.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:211]: done.

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:177]: @ebeworld Can you use the standard vector API?
We are removing the Decimal Vector API in https://github.com/facebookincubator/velox/pull/5571

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:260]: Probably cleaner to add another separate test `decimalToIntegralBounds` with this code under another `testDecimalToIntegralCastBounds` method.

**Review by majetideepak — COMMENTED** [review]
@ebeworld changes look good. Some minor comments.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp:178]: Hoisting computation out of expensive for loop improves performance.

File: velox/expression/tests/CastExprTest.cpp
Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp:184]: Note that expected values are different. This is to test that floating point comparison, where comparison should be within delta.

Comment (karteekmurthys) [reply velox/expression/tests/CastExprTest.cpp:184]: Thanks for explaining that. Would you please add a comment about that.

**Review by karteekmurthys — APPROVED** [review]

File: velox/expression/CastExpr.cpp
Comment (majetideepak) [reply velox/expression/CastExpr.cpp]: @ebeworld can you address this comment?

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:280]: On the Presto side, I see this to be `58`. Can you clarify the semantics?
```
select cast(cast(57.55 as decimal(6, 2)) as integer);
 _col0 
-------
    58 
(1 row)

presto> select cast(decimal '57.55' as BIGINT);
 _col0 
-------
    58 
(1 row)
```

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:231]: Probably use `DECIMAL(10, 0)`
```
explain select decimal '-2147483649';
  ......
Output[_col0] => [expr:decimal(10,0)]  
 ```

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:249]: `DECIMAL(10, 0)`

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:261]: We need to simply the error message. Probably in a follow up PR.
Presto has a simple message
```
select cast(decimal '2147483648' as INT);
Query 20230718_215715_00021_i2xx8 failed: Cannot cast '2147483648' to INTEGER
```

File: velox/expression/CastExpr.cpp
Comment (majetideepak) [inline velox/expression/CastExpr.cpp:249]: This will result in a divide by 0 error if the scaleFactor is 1 or 0.
You use the remainder to roundUp.

**Review by majetideepak — CHANGES_REQUESTED** [review]

File: velox/expression/CastExpr.cpp
Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:249]: May be try to use this function `DecimalUtill::divideWithRoundUp`. https://github.com/facebookincubator/velox/blob/main/velox/type/DecimalUtil.h#L151

File: velox/expression/CastExpr.cpp
Comment (majetideepak) [inline velox/expression/CastExpr.cpp:249]: Can't this be simplified as follows?

```
auto sign = (simpleInput->valueAt(row) > 0) ? 1 : -1;
auto offset = sign * ((scaleFactor > 1) ? (sign * ((scaleFactor >> 1) - 1)) : 0);
auto integralPart = (simpleInput->valueAt(row) +  offset) / scaleFactor;
```

Comment (majetideepak) [reply velox/expression/CastExpr.cpp]: @ebeworld does the above code block not work to get the final result?

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:177]: Is the `type` argument used anywhere?

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:235]: You can get the Type name from the physical type using `TypeTraits<T>::name`. You don't need the Type argument.

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [inline velox/expression/tests/CastExprTest.cpp:266]: `type` not used anywhere?

**Review by majetideepak — COMMENTED** [review]
@ebeworld a couple of more comments. Thanks for clarifying the overflow issue.

File: velox/expression/tests/CastExprTest.cpp
Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp:235]: These are native types, like int32_t,  hence that does not work.

File: velox/expression/tests/CastExprTest.cpp
Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp:266]: Oh, that is left over from previous reviews. Previously, testDecimalToIntegralCastsOutOfBounds test was inside testDecimalToIntegralCasts, hence there was need of type. No longer. Fixed.

File: velox/expression/tests/CastExprTest.cpp
Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp:177]: This is left over from previous reviews. Previously, testDecimalToIntegralCastsOutOfBounds test was inside testDecimalToIntegralCasts, hence there was need of type. No longer. Fixed.

File: velox/expression/tests/CastExprTest.cpp
Comment (majetideepak) [reply velox/expression/tests/CastExprTest.cpp:235]: To clarify, you are passing an argument type only for the name right?

Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp]: Changed into TypeKind and made it work

**Review by majetideepak — APPROVED** [review]
Thanks, @ebeworld

**Comment by majetideepak** [comment]
@mbasmanova this looks ready. Can you make a final pass and help with merging? Thanks.

File: velox/expression/tests/CastExprTest.cpp
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:262]: This message looks wrong. "to {}" should have something other than "{}", no?

**Review by mbasmanova — COMMENTED** [review]
@ebeworld Would you update PR description to provide more details about this change? What it that we had support from decimal to double and now you add support from decimal to real and bigint?

File: velox/expression/CastExpr.cpp
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:210]: We are seeing poor performance of try_cast when many values throw. @laithsakka optimized some of the code paths to avoid throwing. Would it be possible to optimize this code path as well? See https://github.com/facebookincubator/velox/pull/5801

**Review by mbasmanova — COMMENTED** [review]
@kagamiori Wei is adding documentation for CAST in https://github.com/facebookincubator/velox/pull/5847

Let's wait for this to land, then update it to include new casts introduced in this PR.

File: velox/expression/tests/CastExprTest.cpp
Comment (ebeworld) [reply velox/expression/tests/CastExprTest.cpp:262]: The name will replace in placeholder. It will be like Failed to cast from DECIMAL(19,0) to INTEGER: 2147483648. Out of bounds.

**Comment by ebeworld** [comment]
> 

Updated the PR description.

File: velox/expression/CastExpr.cpp
Comment (ebeworld) [reply velox/expression/CastExpr.cpp]: Done.

**Comment by ebeworld** [comment]
Failing benchmark LikeFunctionsBenchmark.cpp/wildcardAtLeastN looks unrelated.

**Comment by mbasmanova** [comment]
> Failing benchmark LikeFunctionsBenchmark.cpp/wildcardAtLeastN looks unrelated.

CC: @pedroerp

File: velox/expression/CastExpr-inl.h
Comment (laithsakka) [inline velox/expression/CastExpr-inl.h:259]: if inside a try_cast, we can set null in the output directly instead of throwing exception here, 
throw+ constructing the exception are expensive, if this happen for each row for some reason it would be 100sX slower.

Comment (laithsakka) [reply velox/expression/CastExpr-inl.h:259]: check void CastExpr::applyCastKernel

if (inputRowValue.size() == 0) {
        if (setNullInResultAtError()) {
          result->setNull(row, true);
        } else {
          context.setVeloxExceptionError(
              row,
              makeBadCastException(
                  result->type(), *input, row, "Empty string"));
        }
        return;
      }

Comment (ebeworld) [reply velox/expression/CastExpr-inl.h]: Thanks @laithsakka. I didn't know there was such mechanism. Changed accordingly.

**Comment by ebeworld** [comment]
@pedroerp @mbasmanova Looks the benchmark failures are not related to the change.

**Comment by pedroerp** [comment]
> @pedroerp @mbasmanova Looks the benchmark failures are not related to the change.

Yes, it's unrelated. We disabled the benchmark signal, so the failure should go away if you rebase it.

**Comment by ebeworld** [comment]
@mbasmanova Rebased. Would you take a look?

File: velox/expression/CastExpr-inl.h
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:269]: I wonder why don't we have 'return' statement in the 'if' block? Do we have tests that cover all the code paths here?

Comment (ebeworld) [reply velox/expression/CastExpr-inl.h:269]: You are right. I just added a test for that path.

Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:269]: For some reason I still see 'return' in 'else' branch, but not in 'if' branch. Should the return be outside of if-else block?

Comment (ebeworld) [reply velox/expression/CastExpr-inl.h:269]: Yes, that is by design.
-  If not throwing and setting null, then setting null for all error cases(tested by testDecimalToIntegralCastsOutOfBoundsSetNullOnFailure where all out of bound errors are set into std::nullopt, expecting no exception). 
- In else case, setting up to throw the first exception is enough (tested by testDecimalToIntegralCastsOutOfBounds, where it is expecting to catch the first throw).

Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:269]: I'm still confused. Why do we need to set resultBuffer[row] when setNullInResultAtError() is true?

Comment (ebeworld) [reply velox/expression/CastExpr-inl.h:269]: My understanding is that applying CAST operator to values equals applying CAST on each, unless it throws. In cases of TRY_CAST like functions(https://learn.microsoft.com/en-us/sql/t-sql/functions/try-convert-transact-sql?view=sql-server-ver16), it does not throw, instead sets corresponding values to null.

In short, TRY_CAST("int", X) where X={'a', '1'} would return {NULL, 1}.  

Do i have misunderstanding?

Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:269]: This understanding is correct. 

The 'return' statement in question is for the per-row lambda passed to applyToSelectedNoThrowLocal. Returning from that lambda doesn't prevent next row from being evaluated, but does prevent the code after the 'return' from executing.

Here, I see that we have a single line of code after if-else:

`resultBuffer[row] = static_cast<To>(integralPart);`

I also see that 'return' in the 'else' block prevents this line from executing, but there is no 'return' in the 'if' block to prevent this line from executing. This is confusing to me. I'd expect a 'return' after an if-else block since we don't need to evaluate `resultBuffer[row] = static_cast<To>(integralPart);` in either case.

Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:269]: Would you do a couple of experiments: (1) remove 'return' statement; (2) move it after the if-else block. Do you observe any change in behavior ( I expect not. )

Comment (ebeworld) [reply velox/expression/CastExpr-inl.h]: Ah, i see. You are right, if null bit is set, it does not need to set the value itself.  I have moved 'return' after if-else. 

1. Removing would cause to set exception multiple times in case of X={'a', 'b', 'c'} which is wasteful. If throwing, the recording the first may be enough. Also setting might override exceptions, so will throw exception about 'c', while throwing exception about 'a' is desirable.

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Comment by mbasmanova** [comment]
PR #5847 added documentation for CAST. Let's rebase and update that documentation to include newly added functionality.

**Comment by mbasmanova** [comment]
@ebeworld Thank you for rebasing. Would you also update the documentation? 

https://facebookincubator.github.io/velox/functions/presto/conversion.html

**Comment by ebeworld** [comment]
@mbasmanova Updated the docs.

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Pull request closed without merging**
